File: 238. Product of Array Except Self.cpp
class Solution2 {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
    	int len = nums.size();
    	if (len == 0)	return vector<int>();
        vector<int> rst(len, 1);
        rst[len - 1] = nums[len - 1];
        rst[0] = nums[0];
        for(int i = len - 2; i > 0; --i)
        	rst[i] = nums[i] * rst[i + 1];
        int last = 1;
        for(int i = 0; i < len - 1; ++i){
        	rst[i] = last * rst[i + 1];
        	last = last * nums[i];
        }
        rst[len - 1] = lst;
        return rst;
    }
};

File: 238. Product of Array Except Self.py
    class Solution(object):

        def productExceptSelf(self, nums):

            """

            :type nums: List[int]

            :rtype: List[int]

            """

            n = len(nums)

            left_product = [1] * n

            right_product = [1] * n

            res = [1] * n

            for i in range(1, n):

                left_product[i] = left_product[i-1]*nums[i-1]

            for i in range(n-2, -1, -1):

                right_product[i] = right_product[i+1]*nums[i+1]

            for i in range(n):

                res[i] = left_product[i]*right_product[i]

            return res

             
